File: Trees/543. Diameter of Binary Tree/Solution.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # N : no of nodes in a tree
        # Time Complexity: O(N)
        # Space Complexity: O(N) -> Stack Space
        total = 0
        def dfs(root):
            nonlocal total
            if not root:
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)
            total = max(total, (left + right))
            return 1 + max(left, right)
        dfs(root)
        return total

File: Trees/543. Diameter of Binary Tree/Solution.java
import java.lang.Math;

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    private int dfs(TreeNode root, int[] diameter) {
        if (root == null) return 0;

        int left = dfs(root.left, diameter);
        int right = dfs(root.right, diameter);
        diameter[0] = Math.max(diameter[0], (left + right));
        return 1 + Math.max(left, right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        // N : no of nodes in a tree
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        int[] diameter = new int[1];
        dfs(root, diameter);
        return diameter[0];
    }
}